core/appmodule/event.go

```diff
@@ -0,0 +1,37 @@
+package appmodule
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+// HasEventListeners is the extension interface that modules should implement to register
+// event listeners.
+type HasEventListeners interface {
+	AppModule
+
+	// RegisterEventListeners registers the module's events listeners.
+	RegisterEventListeners(registrar *EventListenerRegistrar)
+}
+
+// EventListenerRegistrar allows registering event listeners.
+type EventListenerRegistrar struct {
+	listeners []any
+}
+
+// GetListeners gets the event listeners that have been registered
+func (e *EventListenerRegistrar) GetListeners() []any {
+	return e.listeners
+}
```

Comment:
What does this actually do though? Do we know what listeners will look like? How will they subscribe and then handle events?

Reply:
Reflection will still be needed in runtime. It just provides a generic API which should be nicer for users

Reply:
I get that, but it doesn't answer my question. Why are we not discussing how/where these will be listened for? Should we not define that listening interface? Have we considered using some industry standard event bus/listening tool?

Reply:
Well that's a detail for runtime to implement. The idea with this API is just to specify some way for modules to listen and they shouldn't care how routing is done.

Also, we need to emit events in a deterministic order so we probably can't use some off-the-shelf event bus. But it shouldn't be too hard - it's just looking up the list of listeners for an event type in a dictionary. Also, we probably want runtime to be able to specify the order of module per event type, for example for `cosmos.bank.v1.EventValidatorCreated` an app might want to specify the order slashing, distribution. But the default order might be alphabetical (i.e. distribution, slashing).

Reply:
I'm fine with this being an `any` for now.